namer/consul/src/main/scala/io/buoyant/namer/consul/LookupCache.scala

```diff
@@ -33,60 +38,94 @@ private[consul] class LookupCache(
 
-          Closable.make { deadline =>
-            closableFuture.flatMap(_.close(deadline))
+  private[this] val lookupAddress: (Path, String, SvcKey) => Future[InstrumentedVar[Addr]] =
+    (raw, dc, key) => {
+      val pollState = SvcAddr.mkConsulPollState
+      val addrFuture: Future[InstrumentedVar[Addr]] = resolveDc(dc).join(domain).map {
+        case ((dcName, domainOption)) =>
+          SvcAddr(
+            consulApi,
+            LookupCache.DefaultBackoffs,
+            dcName,
+            key,
+            domainOption,
+            consistency = consistency,
+            preferServiceAddress = preferServiceAddress,
+            weights,
+            serviceStats,
+            pollState
+          )
+      }
+
+      //if the lookup succeeded we want to cache the result, double checking for race conditions
+      addrFuture.map {
+        addr =>
+          val cachedAddr = Var(addr)
+          lookupStatusMu.synchronized {
+            if (lookupStatus.containsKey(raw)) {
+              cachedAddr() = lookupStatus.get(raw).addr
+            } else {
+              lookupStatus.put(raw, InstrumentedAddr(addr, pollState))
+              cachedCounter.incr()
+            }
           }
+          cachedAddr()
+      }
+    }
+
+  private[this] val convertToName: (Future[InstrumentedVar[Addr]], Path, Path) => Activity[NameTree[Name]] =
+    (addrFuture, id, residual) => {
+      val observation = Var.async[Activity.State[NameTree[Name.Bound]]](Activity.Pending) { observationState =>
+        val closableFuture = addrFuture.transform {
+          case Return(addr) =>
+            val observationClosable = addr.underlying.changes.respond {
+              case Addr.Neg => observationState.update(Activity.Ok(NameTree.Neg))
+              case Addr.Pending => observationState.update(Activity.Pending)
+              case Addr.Failed(why) => observationState.update(Activity.Failed(why))
+              case Addr.Bound(_, _) => observationState.update(Activity.Ok(NameTree.Leaf(Name.Bound(addr.underlying, id, residual))))
+            }
+            Future.value(observationClosable)
+          case Throw(cause) =>
+            // We probably failed to fetch agent config. This is critical.
+            // TODO: if this has happened only restart can restore consul namer to working state. Throw exception instead?
+            observationState.update(Activity.Failed(cause))
+            Future.value(Closable.nop)
+        }
+
+        Closable.make { deadline =>
+          closableFuture.flatMap(_.close(deadline))
         }
+      }
+      Activity(observation)
+    }
 
-        cachedCounter.incr()
-        new Activity(observation)
+  private[this] val compute: (Path, String, SvcKey, Path, Path) => Activity[NameTree[Name]] =
+    untupled(Memoize[(Path, String, SvcKey, Path, Path), Activity[NameTree[Name]]] {
+      case (raw, dc, key, id, residual) => {
+        val cached = Option(lookupStatus.get(raw))
```

Comment:
This cache within a cache is confusing to reason about.  I think this lookup will always fail because lookupAddress will insert this item into cache.

Reply:
I didn't actually trust enough `Memoize` to avoid race conditions
I've reviewed the source code and I'll remove this redundant check